peachjam/timeline_email_service.py

```diff
@@ -180,58 +175,59 @@ def send_new_citation_email(user):
+                        whole_work=False,
+                    )[:2]
+                    citing_documents.append(
+                        {
+                            "document": doc,
+                            "provision_citations": provision_citations,
+                        }
+                    )
+
+                context["saved_documents"].append(
                     {
-                        "document": doc,
-                        "provision_citations": provision_citations,
+                        "saved_document": ev.user_following.saved_document.document,
+                        "citing_documents": citing_documents,
                     }
                 )
+            site = Site.objects.get_current()
+            context["site_domain"] = f"https://{site.domain}"
 
-            context["saved_documents"].append(
-                {
-                    "saved_document": ev.user_following.saved_document.document,
-                    "citing_documents": citing_documents,
-                }
+            # render html template string
+            context["html_body"] = render_to_string(
+                "peachjam/emails/new_citation_alert_body.html", context=context
             )
-        site = Site.objects.get_current()
-        context["site_domain"] = f"https://{site.domain}"
+            subject_line = f"New citations for {context['saved_documents'][0]['saved_document'].title}"
+            saved_docs_length = len(context["saved_documents"])
+            if saved_docs_length > 1:
+                subject_line += f" and {saved_docs_length} more"
 
-        # render html template string
-        context["html_body"] = render_to_string(
-            "peachjam/emails/new_citation_alert_body.html", context=context
-        )
-        subject_line = (
-            f"New citations for {context['saved_documents'][0]['saved_document'].title}"
-        )
-        saved_docs_length = len(context["saved_documents"])
-        if saved_docs_length > 1:
-            subject_line += f" and {saved_docs_length} more"
-
-        context["subject_line"] = subject_line
-
-        with override(user.userprofile.preferred_language.pk):
-            send_templated_mail(
-                template_name="new_citation_alert",
-                from_email=settings.DEFAULT_FROM_EMAIL,
-                recipient_list=[user.email],
-                context=context,
-            )
+            context["subject_line"] = subject_line
 
-            for ev in events:
-                ev.mark_as_sent()
+            with override(user.userprofile.preferred_language.pk):
+                send_templated_mail(
+                    template_name="new_citation_alert",
+                    from_email=settings.DEFAULT_FROM_EMAIL,
+                    recipient_list=[user.email],
+                    context=context,
+                )
+
+        for ev in events:
+            ev.mark_as_sent()
```

Comment:
Kind of a Nit: But in the event of an email service failure/crash halfway through doing it's thing, the email maybe sent, but it won't be marked. So duplicate emails will still occur. 
Maybe using guaranteed atomicity via transaction.atomic() for future feature improvements?

Reply:
`send_templated_email` isn't asynchronous. Once that API succeeds, the code will continue and call `mark_as_sent()`. If the API call fails, it will raise an exception and so `mark_as_sent()` won't be reached. 

If the email API returns success but can't (for some reason) send the email, then that's out of our hands.